core/arch/x86/mangle.c

```diff
@@ -2391,6 +2394,31 @@ mangle_interrupt(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
 }
 
 /***************************************************************************
+ * Single step exceptions generation
+ */
+void
+mangle_single_step(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr)
+{
+    /*
+     * Simply inserts a nop so as that next instruction where a single step
+     * exception might occur is not in a different basic block.
+     * FIXME i#2144 : to be absolutely transparent, we should translate the
+     * exception address as if we did not insert this nop.
+     */
+    instr_t* next = instr_get_next_app(instr);
+    /* Assumes popf cannot end a basic block. */
+    ASSERT(next);
+    /*
+     * The single step exception is only a problem on a control transfer
+     * because the ExceptionAddress should be the jump target.
+     */
+    if (instr_is_cti(next)) {
+        /* Next app instruction gives a correct translation to this nop. */
```

Comment:
Yes.  Right.  I also looked at the manual and now I believe I finally have all the details.  Sorry, I should have thought this through earlier.  Unfortunately this nop solution is not going to work for the general case.  We can make it work for "popf;jmp" or "popf;jcc" with no client, because we can emulate the jmp or jcc, but the presence of instrumentation means we need a different solution as we can't easily emulate arbitrary sequences of instructions.  I believe we have to disable and later forge the single-step exception.  I will add something to the issue tracker.